### PR TITLE
Add 0.8 migration

### PR DESCRIPTION
**Pull request by marktoda**
Migrate to 0.8.15, on a new branch `0.8`. One setback is that `NFTDescriptor` was failing with an assembly stacktoodeep issue that I couldn't easily fix. Something to do with the new yul optimizer pipeline in 0.8 I think. I shimmed it out, just returning "" for tokenURI for now with comments to add back in at later point - this PR gets us 90% of the way there

### CODE REVIEW

File: test/__snapshots__/LiquidityAmounts.spec.ts.snap [hunk 1]
```diff
@@ -1,21 +1,21 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
-exports[`LiquidityAmounts #getAmount0ForLiquidity gas 1`] = `352`;
+exports[`LiquidityAmounts #getAmount0ForLiquidity gas 1`] = `479`;
```
Comment (willpote) [inline test/__snapshots__/LiquidityAmounts.spec.ts.snap:3]: Any idea why this is going up? This library is used when adding liquidity so best if it didn't change too much

File: test/__snapshots__/PoolAddress.spec.ts.snap [hunk 2]
```diff
@@ -1,5 +1,5 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
-exports[`PoolAddress #computeAddress gas cost 1`] = `642`;
+exports[`PoolAddress #computeAddress gas cost 1`] = `643`;
 
-exports[`PoolAddress #computeAddress matches example from core repo 1`] = `"0x03D8bab195A5BC23d249693F53dfA0e358F2650D"`;
+exports[`PoolAddress #computeAddress matches example from core repo 1`] = `"0xcf75931a4C18d87F784a91719DefFDdb74b7B0d5"`;
```
Comment (willpote) [inline test/__snapshots__/PoolAddress.spec.ts.snap:5]: Don't think this should change?

File: hardhat.config.ts [hunk 3]
```diff
@@ -19,12 +18,11 @@ const LOW_OPTIMIZER_COMPILER_SETTINGS = {
...
 
 const LOWEST_OPTIMIZER_COMPILER_SETTINGS = {
-  version: '0.7.6',
+  version: '0.8.13',
   settings: {
-    evmVersion: 'istanbul',
     optimizer: {
       enabled: true,
-      runs: 1_000,
+      runs: 100,
```
Comment (willpote) [inline hardhat.config.ts:25]: Why lower?

File: foundry.toml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+[default]
```
Comment (willpote) [inline foundry.toml:1]: Used?

File: test/__snapshots__/Base64.spec.ts.snap [hunk 5]
```diff
@@ -1,32 +1,32 @@
...
+exports[`Base64 #encode gas cost of encode(test string) 1`] = `1656`;
 
-exports[`Base64 #encode gas cost of encode(this is a test) 1`] = `1738`;
+exports[`Base64 #encode gas cost of encode(this is a test) 1`] = `1877`;
 
-exports[`Base64 #encode gas cost of encode(this is a very long string that should cost a lot of gas to encode :)) 1`] = `6083`;
+exports[`Base64 #encode gas cost of encode(this is a very long string that should cost a lot of gas to encode :)) 1`] = `5862`;
 
-exports[`Base64 #encode gas cost of encode(😀) 1`] = `1004`;
+exports[`Base64 #encode gas cost of encode(😀) 1`] = `1203`;
```
Comment (willpote) [inline test/__snapshots__/Base64.spec.ts.snap:30]: Guessing this is all due to external library?

**Review by willpote — COMMENTED** [review]
I don't think we should merge until we resolve the NFT descriptor issue, I'd be worried that people may use the 0.8 branch to deploy the code to new networks, which is happening a lot more frequently recently (e.g. Celo/Moonbeam/Gnosis soon)

**Comment by marktoda** [comment]
> I don't think we should merge until we resolve the NFT descriptor issue

Agreed - was able to find a workaround. Apparently ABIEncoderv2 causes stacktoodeep on 0.8 with large structs pretty commonly: https://github.com/ethereum/solidity/issues/12343. Just broke the struct in half

File: contracts/libraries/PoolAddress.sol [hunk 6]
```diff
@@ -3,7 +3,7 @@ pragma solidity >=0.5.0;
 
 /// @title Provides functions for deriving a pool address from the factory, tokens, and the fee
 library PoolAddress {
-    bytes32 internal constant POOL_INIT_CODE_HASH = 0xe34f199b19b2b4f47f68442619d555527d244f78a3297ea89325f843f87b8b54;
+    bytes32 internal constant POOL_INIT_CODE_HASH = 0xa598dd2fba360510c5a8f02f44423a4468e902df5857dbce3ca162a43a3a31ff;
```
[the remark this answers is not in the document]
Comment (marktoda) [reply contracts/libraries/PoolAddress.sol:6]: We're targeting the 0.8 branch of v3-core now, in which the bytecode has changed

File: test/__snapshots__/LiquidityAmounts.spec.ts.snap (hunk 1 above)
Comment (marktoda) [reply test/__snapshots__/LiquidityAmounts.spec.ts.snap:3]: I think unchecked -> checked math. let me see if I can improve

Comment (marktoda) [reply test/__snapshots__/LiquidityAmounts.spec.ts.snap:3]: added a few unchecked blocks and much closer now

File: foundry.toml (hunk 4 above)
Comment (marktoda) [reply foundry.toml:1]: no I suppose not :(

File: test/__snapshots__/Base64.spec.ts.snap (hunk 5 above)
Comment (marktoda) [reply test/__snapshots__/Base64.spec.ts.snap:30]: yeah

File: test/__snapshots__/PoolAddress.spec.ts.snap (hunk 2 above)
Comment (marktoda) [reply test/__snapshots__/PoolAddress.spec.ts.snap:5]: The initcode hash changed for the 0.8 migration in core, so the create2 addresses change

Comment (marktoda) [reply test/__snapshots__/PoolAddress.spec.ts.snap:5]: Added some logs to core-v3#0.8 to test:
```
bytecodeHash 0xa598dd2fba360510c5a8f02f44423a4468e902df5857dbce3ca162a43a3a31ff
factory 0x5FbDB2315678afecb367f032d93F642f64180aa3
tokens [
  '0x1000000000000000000000000000000000000000',
  '0x2000000000000000000000000000000000000000'
]
feeAmount 250
bytecodeHash 250
create2Address 0xcf75931a4C18d87F784a91719DefFDdb74b7B0d5
      ✔ enables pool creation (126ms)

```

File: test/NFTDescriptor.spec.ts [hunk 7]
```diff
@@ -1,907 +0,0 @@
-import { BigNumber, constants, Wallet } from 'ethers'
```
Comment (willpote) [inline test/NFTDescriptor.spec.ts:1]: Same

File: test/__snapshots__/TickLens.spec.ts.snap [hunk 8]
```diff
@@ -1,5 +1,5 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
-exports[`TickLens #getPopulatedTicksInWord fully populated ticks 1`] = `2803889`;
+exports[`TickLens #getPopulatedTicksInWord fully populated ticks 1`] = `2959038`;
 
-exports[`TickLens #getPopulatedTicksInWord gas for single populated tick 1`] = `55664`;
+exports[`TickLens #getPopulatedTicksInWord gas for single populated tick 1`] = `90708`;
```
Comment (willpote) [inline test/__snapshots__/TickLens.spec.ts.snap:5]: Kinda huge increase. TickLens isn't meant to be used on-chain so not a huge deal, but still worth investigating

File: contracts/test/NFTDescriptorTest.sol [hunk 9]
```diff
@@ -1,75 +0,0 @@
-// SPDX-License-Identifier: UNLICENSED
```
[the remark this answers is not in the document]
Comment (marktoda) [reply contracts/test/NFTDescriptorTest.sol:1]: oh oops, forgot to add back i suppose

File: package.json [hunk 10]
```diff
@@ -20,7 +20,7 @@
     "contracts/base",
     "contracts/interfaces",
     "contracts/libraries",
-    "artifacts/contracts/**/*.json",
+    "artifacts/contracts/se**/*.json",
```
[the remark this answers is not in the document]
Comment (marktoda) [reply package.json:23]: oops

File: test/NonfungiblePositionManager.spec.ts [hunk 11]
```diff
@@ -1155,11 +1155,11 @@ describe('NonfungiblePositionManager', () => {
       await expect(nft.tokenURI(tokenId + 1)).to.be.reverted
     })
 
-    it('returns a data URI with correct mime type', async () => {
+    xit('returns a data URI with correct mime type', async () => {
       expect(await nft.tokenURI(tokenId)).to.match(/data:application\/json;base64,.+/)
     })
 
-    it('content is valid JSON and structure', async () => {
+    xit('content is valid JSON and structure', async () => {
```
[the remark this answers is not in the document]
Comment (marktoda) [reply test/NonfungiblePositionManager.spec.ts:1162]: oops, removed

**Review by willpote — APPROVED** [review]

**Comment by nnnnicholas** [comment]
@willpote @marktoda Is this ready branch in a finished and production ready state? Is there any reason it is not yet merged to main? Thank you!

**Comment by ghost** [comment]
!

**Pull request merged**